### ISSUE TITLE
Destructuring with initializer unexpectedly collapses union type 

### ISSUE THREAD

**Issue by jwbay** [Microsoft/TypeScript#22338]
**TypeScript Version:**  typescript@2.8.0-dev.20180302

**Search Terms:** 
narrow destructuring initializer

**Code**

```ts
type Fn = () => number;

interface Foo {
    bar: {} | Fn;
}

function test(params: Foo) {
    const { bar = (() => 42) } = params; // type of bar is {}
    let result: number;

    if (typeof bar === 'function') {
        result = bar(); // type is never, unexpected error
    }

    if (typeof params.bar === 'function') {
        result = params.bar();  // ok
    }

    const { bar: noInitializer } = params;
    if (typeof noInitializer === 'function') {
        result = noInitializer(); // ok
    }
}
```

**Expected behavior:**
All assignments to result are ok

**Actual behavior:**
First assignment fails. The presence of the initializer seems to collapse the function type into {} since it's assignable to it. This is undesired though, especially given the other two cases which work.

**Playground Link:** [link](https://www.typescriptlang.org/play/#src=type%20Fn%20%3D%20()%20%3D%3E%20number%3B%0D%0A%0D%0Ainterface%20Foo%20%7B%0D%0A%20%20%20%20bar%3A%20%7B%7D%20%7C%20Fn%3B%0D%0A%7D%0D%0A%0D%0Afunction%20test(params%3A%20Foo)%20%7B%0D%0A%20%20%20%20const%20%7B%20bar%20%3D%20(()%20%3D%3E%2042)%20%7D%20%3D%20params%3B%20%2F%2F%20type%20of%20bar%20is%20%7B%7D%0D%0A%20%20%20%20let%20result%3A%20number%3B%0D%0A%0D%0A%20%20%20%20if%20(typeof%20bar%20%3D%3D%3D%20'function')%20%7B%0D%0A%20%20%20%20%20%20%20%20result%20%3D%20bar()%3B%20%2F%2F%20type%20is%20never%2C%20unexpected%20error%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20if%20(typeof%20params.bar%20%3D%3D%3D%20'function')%20%7B%0D%0A%20%20%20%20%20%20%20%20result%20%3D%20params.bar()%3B%20%20%2F%2F%20ok%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20const%20%7B%20bar%3A%20noInitializer%20%7D%20%3D%20params%3B%0D%0A%20%20%20%20if%20(typeof%20noInitializer%20%3D%3D%3D%20'function')%20%7B%0D%0A%20%20%20%20%20%20%20%20result%20%3D%20noInitializer()%3B%20%2F%2F%20ok%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A)

**Related Issues:**
https://github.com/Microsoft/TypeScript/issues/14721, but that seems to be fixed

**Comment by mhegazy** [comment]
`{}` is a super type of all functions. so `{} | fn` is just `{}`.

**Comment by jwbay** [comment]
Seems like this wouldn't work, then?
```ts
let x: {} |(() => number);

if (typeof x === 'function') {
    // {} is gone from x here
    x();
}
```

**Comment by jwbay** [comment]
^ I think that example is not super relevant, the problem is the type from the destructure. More weirdness:
 
```ts
interface Base {
    x: number;
}

interface Super extends Base {
    y: string;
}

let foo: { bar: Base | Super }

const {
    bar = { x: 42, y: '' } // bar is Base, and providing y is an error. wat?
} = foo;

const {
    bar, // bar is Base | Super 👌
} = foo;
```

**Comment by mhegazy** [comment]
The subtypes in a union are not aggressively reduced. they are only reduced when the type is needed for a declaration, e.g. the return type of a function inferred from multiple return statements, or the type of a variable inferred from an intializer. in other cases you can witness the union type that is reducible. These are all implementation details, and could change. but the fact remains that `{} | someOtherType` is just `{}`. similarly `Base | Derived` is just `Base`. the TypeScript type system is a structural one, and thus a more general type in a union will subsume the more special one.

**Comment by jwbay** [comment]
Thanks for the explanation! Still seems like a bug to me that I can't initialize with one of the declared types, but ¯\_(ツ)_/¯
